### PR TITLE
Add Bedrock LLM

### PR DESCRIPTION
**Pull request by neil-botelho**
# Description
This commit adds support for bedrock LLMs. I've added unit tests for basic and streaming functionality.  A couple of notes on some design decisions I've made that you may want done differently:

1. the Anthropic model requires the prompt to start with "Human:" and end with "Assistant:". But boto3 gives very detailed error messages for these cases. So I don't raise an error if `complete` is called when using an Anthropic model, I allow boto3 to raise the error. 
1. In Bedrock.metadata I've set is_chat_model to be False, based on it's definition in the LLMMetadata class. All calls to bedrock models take a string prompt as input, they don't accept a series of messages. But some models(eg. Anthropic) are technically chat models. 

Let me know if either of these are wrong, I'm happy to update the PR

## Type of Change

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

# How Has This Been Tested?

- [x] Added new unit/integration tests
- [x] Added new notebook (that tests end-to-end)
- [x] I stared at the code and made sure it makes sense

# Suggested Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I ran `make format; make lint` to appease the lint gods

### CODE REVIEW

File: llama_index/llms/bedrock.py [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+    def stream_chat(
+        self, messages: Sequence[ChatMessage], **kwargs: Any
+    ) -> ChatResponseGen:
+        stream_chat_fn = stream_completion_to_chat_decorator(self.stream_complete)
+        return stream_chat_fn(messages, **kwargs)
+
+    async def achat(
+        self, messages: Sequence[ChatMessage], **kwargs: Any
+    ) -> ChatResponse:
+        raise NotImplementedError
```
Comment (hatianzhang) [inline llama_index/llms/bedrock.py:210]: thanks @neil-botelho for the contribution. 
I am wondering do you have any plans to add those callback functions?
thanks

Comment (neil-botelho) [reply llama_index/llms/bedrock.py:210]: By callback functions do you mean the async chat and completion endpoints? As far as I can tell, boto3 doesn't support async so we can't make async calls to bedrock. If I'm mistaken or if bedrock adds support to make async calls I'd be happy to update my pr or add a new pr to add those endpoints
References:
- [this stackoverflow article](https://stackoverflow.com/a/72093772)
- [The boto3 bedrock-runtime docs](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/bedrock-runtime/client/invoke_model.html) lacking an async parameter

File: docs/examples/llm/bedrock.ipynb
Comment (Disiok) [inline docs/examples/llm/bedrock.ipynb:1]: nit: notebook title typo

File: llama_index/llms/bedrock.py (hunk 1 above)
Comment (neil-botelho) [reply llama_index/llms/bedrock.py:210]: I did notice the llama_index.llms.CustomLLM is a subclass of LLM that implements the async chat and completion endpoints by just calling the sync chat and completion endpoints from an async function. As far as I can tell, the call to the LLM will still be blocking in this case. Hence I didn't go for that approach.

File: docs/examples/llm/bedrock.ipynb
Comment (Disiok) [inline docs/examples/llm/bedrock.ipynb:1]: Could you help link this into our docs? See: docs/core_modules/model_modules/llms/modules.md

File: llama_index/llms/bedrock.py [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+from llama_index.llms.custom import CustomLLM
+
+
+class Bedrock(LLM):
+    model: str = Field(description="The modelId of the Bedrock model to use.")
+    temperature: float = Field(description="The temperature to use for sampling.")
+    max_tokens: int = Field(description="The maximum number of tokens to generate.")
+    context_size: int = Field("The maximum number of tokens available for input.")
+    profile_name: Optional[str] = Field(
+        description="he name of aws profile to use. If not given, then the default profile is used."
```
Comment (Disiok) [inline llama_index/llms/bedrock.py:39]: very nit: typo on "(t)he name"

File: llama_index/llms/bedrock.py [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+            raise ValueError(
+                "`context_size` argument not provided and"
+                "model provided refers to a non-foundation model."
+                " Please specify the context_size"
+            )
+        try:
+            import boto3
+
+            aws_access_key_id
+            aws_secret_access_key
```
Comment (Disiok) [inline llama_index/llms/bedrock.py:82]: nit: remove?

**Review by Disiok — APPROVED** [review]
Amazing! Thanks so much for the contribution!

File: llama_index/llms/bedrock.py (hunk 1 above)
Comment (Disiok) [reply llama_index/llms/bedrock.py:210]: Yep that makes sense, thanks for the explanation!

File: docs/examples/llm/bedrock.ipynb
Comment (neil-botelho) [reply docs/examples/llm/bedrock.ipynb:1]: Sure! Is there any particular order followed here?

Comment (Disiok) [reply docs/examples/llm/bedrock.ipynb:1]: nope, ordered by when integration is introduced mostly

Comment (neil-botelho) [reply docs/examples/llm/bedrock.ipynb:1]: so Bedrock would be added at the end then. Got it.

**Review by hatianzhang — APPROVED** [review]

**Comment by hatianzhang** [comment]
awesome! thanks!

**Pull request merged**